File: analyze_grid.js
const XLSX = require('xlsx');
const fs = require('fs');
const path = require('path');

const filePath = path.join(__dirname, 'grid_points_businesses_window tinting scarborough.xlsx');

try {
    if (!fs.existsSync(filePath)) {
        console.error(`File not found: ${filePath}`);
        process.exit(1);
    }

    const workbook = XLSX.readFile(filePath);
    const sheetName = workbook.SheetNames[0];
    const sheet = workbook.Sheets[sheetName];
    const data = XLSX.utils.sheet_to_json(sheet);

    console.log(`Total rows: ${data.length}`);
    if (data.length > 0) {
        console.log("First row keys:", Object.keys(data[0]));
        console.log("First row sample:", JSON.stringify(data[0], null, 2));
    }

    // Map to store counts: Business Name -> Count of Top 3 Rankings
    const topRankCounts = {};

    data.forEach(row => {
        // Try to find Rank and Business Name columns flexibly
        let rank = row['Rank'] || row['rank'];
        let businessName = row['Business Name'] || row['business name'] || row['Business'] || row['Name'];

        // If simple lookup failed, try to find keys containing specific words
        if (rank === undefined) {
            const rankKey = Object.keys(row).find(k => k.toLowerCase().includes('rank'));
            if (rankKey) rank = row[rankKey];
        }
        if (businessName === undefined) {
            const nameKey = Object.keys(row).find(k => k.toLowerCase().includes('business') || k.toLowerCase().includes('name'));
            if (nameKey) businessName = row[nameKey];
        }

        const parsedRank = parseInt(rank);

        if (!isNaN(parsedRank) && parsedRank >= 1 && parsedRank <= 3 && businessName) {
            if (!topRankCounts[businessName]) {
                topRankCounts[businessName] = 0;
            }
            topRankCounts[businessName]++;
        }
    });

    // Convert to array and sort
    const sortedCompetitors = Object.entries(topRankCounts)
        .sort((a, b) => b[1] - a[1]);

    console.log("\n--- Analysis Result ---");
    console.log("Top Competitors (Rank 1-3 Count):");
    sortedCompetitors.forEach(([name, count], index) => {
        console.log(`${index + 1}. ${name}: ${count}`);
    });

    if (sortedCompetitors.length > 0) {
        console.log(`\nTop competitor is: ${sortedCompetitors[0][0]}`);
    } else {
        console.log("\nNo competitors found ranking in positions 1-3.");
    }

} catch (error) {
    console.error("Error analyzing file:", error);
}

File: analyze_grid.py
import pandas as pd
import os
import sys

# Set encoding to utf-8 for console output
sys.stdout.reconfigure(encoding='utf-8')

file_path = r'C:\Users\Pemina\Documents\Beyond Detail\WEBSITE\Beyond_Detail\grid_points_businesses_window tinting scarborough.xlsx'

print(f"Analyzing file: {file_path}")

try:
    if not os.path.exists(file_path):
        print("Error: File not found.")
        sys.exit(1)

    # Load the Excel file
    try:
        df = pd.read_excel(file_path)
        print("File loaded successfully.")
    except Exception as e:
        print(f"Error loading Excel file via Pandas: {e}")
        # Fallback or exit?
        sys.exit(1)
    
    # Normalize column names
    df.columns = [str(col).strip() for col in df.columns]
    print(f"Columns found: {list(df.columns)}")

    # Identify relevant columns
    # We look for 'Rank' and 'Business Name'
    rank_col = next((c for c in df.columns if 'rank' in c.lower()), None)
    name_col = next((c for c in df.columns if 'business' in c.lower() or 'name' in c.lower()), None)

    if not rank_col:
        print("Could not find a 'Rank' column.")
    if not name_col:
        print("Could not find a 'Business Name' column.")

    if rank_col and name_col:
        print(f"Using columns: Rank='{rank_col}', Business='{name_col}'")
        
        # Filter for top 3 rankings
        # Ensure rank is numeric
        df[rank_col] = pd.to_numeric(df[rank_col], errors='coerce')
        
        top3_df = df[(df[rank_col] >= 1) & (df[rank_col] <= 3)]
        
        if top3_df.empty:
            print("No top 3 rankings found.")
        else:
            # Count businesses
            counts = top3_df[name_col].value_counts()
            
            print("\n--- TOP COMPETITORS (Rank 1-3) ---")
            for name, count in counts.items():
                print(f"{name}: {count}")
            
            print(f"\nTop Competitor: {counts.index[0]}")
    else:
        print("Showing first 3 rows to help debug:")
        print(df.head(3))

except Exception as e:
    print(f"An unexpected error occurred: {e}")
